crates/interpreter/src/instructions/host.rs

```diff
@@ -103,14 +103,28 @@ pub fn extcodecopy<H: Host + ?Sized, SPEC: Spec>(interpreter: &mut Interpreter,
         .set_data(memory_offset, code_offset, len, &code.original_bytes());
 }
 
-pub fn blockhash<H: Host + ?Sized>(interpreter: &mut Interpreter, host: &mut H) {
+pub fn blockhash<H: Host + ?Sized, SPEC: Spec>(interpreter: &mut Interpreter, host: &mut H) {
     gas!(interpreter, gas::BLOCKHASH);
     pop_top!(interpreter, number);
 
-    if let Some(diff) = host.env().block.number.checked_sub(*number) {
+    let block_number = host.env().block.number;
+
+    if let Some(diff) = block_number.checked_sub(*number) {
         let diff = as_usize_saturated!(diff);
+
         // blockhash should push zero if number is same as current block number.
-        if diff <= BLOCK_HASH_HISTORY && diff != 0 {
+        if SPEC::enabled(PRAGUE) && diff <= HISTORY_SERVE_WINDOW && diff != 0 {
+            let Some((value, is_cold)) = host.sload(
+                HISTORY_STORAGE_ADDRESS,
+                number.wrapping_rem(U256::from(HISTORY_SERVE_WINDOW)),
+            ) else {
+                interpreter.instruction_result = InstructionResult::FatalExternalError;
+                return;
+            };
+            gas!(interpreter, gas::sload_cost(SPEC::SPEC_ID, is_cold));
```

Comment:
Maybe extract the `SLOAD` logic from its function into a separate function where we also call here?

Reply:
yeah i had this in mind too, i just forgor

Reply:
i tried making a func but mut borrow of interpreter from `pop_top!` prevented me, so I made an `sload!` macro instead